python/ray/train/base_trainer.py

```diff
@@ -369,6 +369,7 @@ def _generate_trainable_cls(self) -> Type["Trainable"]:
             A Trainable class to use for training.
         """
 
+        from ray.data.context import DatasetContext
```

Comment:
can we do a safer import here? this forces Dataset dependency on users who may not be using datasets.

we can just gate all this logic under `if has_base_dataset:` https://github.com/ray-project/ray/pull/29192/files#diff-c1b2abe5763b7df45c2720fabaa55cb075f93126f5012ebf551c91d855a34494R399

Reply:
Ah good call! That's an easy way to gate it.